db/version_set.h

```diff
@@ -130,6 +130,37 @@ class VersionStorageInfo {
 
   void AddFile(int level, FileMetaData* f);
 
+  // Resize/Initialize the space for compact_cursor_
+  void ResizeCompactCursors(int level) {
+    compact_cursor_.resize(level, InternalKey());
+  }
+
+  const std::vector<InternalKey>& GetCompactCursors() const {
+    return compact_cursor_;
+  }
+
+  // REQUIRES: ResizeCompactCursors has been called
+  void AddCursorForOneLevel(int level,
+                            const InternalKey& smallest_uncompacted_key) {
+    compact_cursor_[level] = smallest_uncompacted_key;
+  }
+
+  // REQUIRES: lock is held
+  // Update the compact cursor and advance the file index so that it can point
+  // to the next cursor
+  const InternalKey& AdvanceCompactCursor(int level) {
+    int cmp_idx = ++next_file_to_compact_by_size_[level];
+    InternalKey new_cursor;
+    if (cmp_idx >= (int)files_by_compaction_pri_[level].size()) {
+      new_cursor = InternalKey();
```

Comment:
The first file has been picked in the last compaction, and we need to find the next cursor. When picking a file, the compaction_picker_level always starts with `next_file_to_compact_by_size_[level]`, I think we need to update it here. Otherwise, the first one may still be compacted and current implementation just returns false if the pointed file is being compacted.

Reply:
Here I assume some other new files with smaller `smallest keys` may be added into this level. In this case, the next time we want to pick one file, we directly start from the smallest file instead of the first file that has larger or equivalent `smallest keys` to the cursor. If we set using the first file's smallest key, we may skip those new files with smaller `smallest keys`. And until the next round, we cannot get a chance to pick them.

Reply:
Got it. I wonder if  changing it though would give a useful guarantee, like oldest data in the level will be compacted next. I haven't thought through it yet so am fine with this approach for now.